src/impl_/trampoline.rs

```diff
@@ -65,26 +70,34 @@ trampolines!(
 );
 
 #[inline]
-pub unsafe fn getter(
+pub(crate) unsafe extern "C" fn getter(
     slf: *mut ffi::PyObject,
     closure: *mut c_void,
-    f: for<'py> unsafe fn(Python<'py>, *mut ffi::PyObject) -> PyResult<*mut ffi::PyObject>,
 ) -> *mut ffi::PyObject {
-    // PyO3 doesn't use the closure argument at present.
-    debug_assert!(closure.is_null());
-    trampoline_inner(|py| f(py, slf))
+    // Safety: PyO3 sets the closure when constructing the ffi getter so this cast should always be valid
+    let closure: &GetSetDefClosure = &*(closure as *const GetSetDefClosure);
+    // Safety: PyO3 sets the getter field when constructing the ffi getter
+    let getter = match closure.getter {
+        Some(getter) => getter,
+        None => std::hint::unreachable_unchecked(),
+    };
+    trampoline_inner(|py| getter(py, slf))
 }
 
 #[inline]
-pub unsafe fn setter(
+pub(crate) unsafe extern "C" fn setter(
     slf: *mut ffi::PyObject,
     value: *mut ffi::PyObject,
     closure: *mut c_void,
-    f: for<'py> unsafe fn(Python<'py>, *mut ffi::PyObject, *mut ffi::PyObject) -> PyResult<c_int>,
 ) -> c_int {
-    // PyO3 doesn't use the closure argument at present.
-    debug_assert!(closure.is_null());
-    trampoline_inner(|py| f(py, slf, value))
+    // Safety: PyO3 sets the closure when constructing the ffi setter so this cast should always be valid
+    let closure: &GetSetDefClosure = &*(closure as *const GetSetDefClosure);
+    // Safety: PyO3 sets the setter field when constructing the ffi setter
+    let setter = match closure.setter {
+        Some(setter) => setter,
+        None => std::hint::unreachable_unchecked(),
```

Comment:
I am not happy about the `unreachable_unchecked` here. This is because we use `.or_default()` when constructing setter and getter one after the other, right? Is there no way to avoid these effectively invalid instances of `GetSetDefClosure`? Separate `HashMap`s for setters and getters?

Reply:
Agreed.  We can only put a single `*closure` into the `ffi::PyGetSetDef`, but we could encode the three cases with different structs instead of optionality. (Would need a getter-only, setter-only, and both, for a total of 4 trampolines instead of 2.)

I'll push a refactor to achieve that in the coming days.